File: 0079-word-search/0079-word-search.java
class Solution {
    private boolean[][] visited;
    private String word;
    private char[][] board;

    private boolean check(int i, int j, int idx){
        if(idx == word.length()){
            return true;
        }

        if(i<0 || i>=board.length || 
            j<0 || j>=board[0].length ||
            board[i][j] != word.charAt(idx) || visited[i][j]){
                return false;
            }
        
        visited[i][j] = true;
        if(check(i-1, j, idx+1) ||
            check(i+1, j, idx+1) ||
            check(i, j-1, idx+1) ||
            check(i, j+1, idx+1)){
                return true;
            }

        visited[i][j] = false;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.word = word;
        visited = new boolean[board.length][board[0].length];

        for(int i=0;i<board.length; i++){
            for(int j=0;j<board[0].length;j++){
                if(board[i][j] == word.charAt(0) && check(i, j, 0)){
                    return true;
                }
            }
        }
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        m = len(board)
        n = len(board[0])
        visited = [[False] * n for _ in range(m)]
        
        def dfs(i, j, char_idx):
            if char_idx == len(word):
                return True

            if i < 0 or i >= m or j < 0 or j >= n:
                return False
            if visited[i][j] or board[i][j] != word[char_idx]:
                return False

            visited[i][j] = True

            for dx, dy in [[0,1], [1,0], [-1,0], [0,-1]]:
                r = i + dx 
                c = j + dy

                if dfs(r, c, char_idx+1):
                    return True

            visited[i][j] = False
            return False

        for i in range(0, m):
            for j in range(0, n):
                if dfs(i, j, 0):
                    return True

        return False